src/zjs_dgram.c

```diff
@@ -0,0 +1,334 @@
+    CHECK(net_context_get(family, SOCK_DGRAM, IPPROTO_UDP, &udp_sock));
+
+    jerry_value_t sockobj = jerry_create_object();
+    jerry_set_prototype(sockobj, zjs_dgram_socket_prototype);
+
+    dgram_handle_t *handle = zjs_malloc(sizeof(handle));
+    if (!handle)
+        return zjs_error("createSocket: OOM");
+    handle->udp_sock = udp_sock;
+    handle->message_cb_id = -1;
+
+    jerry_set_object_native_handle(sockobj, (uintptr_t)handle, zjs_dgram_free_cb);
+
+    // Can't call this here due to bug in Zephyr - called in .bind() instead
+    //CHECK(net_context_recv(udp_sock, udp_received, K_NO_WAIT, handle));
+
+    return sockobj;
+}
+
+static jerry_value_t zjs_dgram_sock_on(const jerry_value_t function_obj,
+                                    const jerry_value_t this,
+                                    const jerry_value_t argv[],
+                                    const jerry_length_t argc)
+{
+    if (argc < 2 ||
+        !jerry_value_is_string(argv[0]) ||
+        (!jerry_value_is_function(argv[1]) &&
+         !jerry_value_is_null(argv[1]))) {
+        return zjs_error("zjs_dgram_sock_on: invalid argument");
+    }
+
+    GET_HANDLE(dgram_handle_t, handle);
+
+    jerry_size_t str_sz = 32;
+    char event[str_sz];
+    zjs_copy_jstring(argv[0], event, &str_sz);
+    if (strcmp(event, "message"))
+        return zjs_error("zjs_dgram_sock_on: unsupported event type");
+
+    if (handle->message_cb_id != -1)
+        zjs_remove_callback(handle->message_cb_id);
+    if (!jerry_value_is_null(argv[1]))
+        handle->message_cb_id = zjs_add_callback(argv[1], this, handle, NULL);
+
+    return ZJS_UNDEFINED;
+}
+
+// Zephyr "packet sent" callback
+static void udp_sent(struct net_context *context, int status, void *token,
+                     void *user_data)
+{
+    DBG_PRINT("udp_sent: %p, st=%d udata=%p\n", context, status, user_data);
+
+    if (user_data) {
+        jerry_value_t args[1] = {ZJS_UNDEFINED};
+        if (status != 0) {
+            char errbuf[8];
+            snprintf(errbuf, sizeof(errbuf), "%d", status);
+            args[0] = zjs_standard_error(NetworkError, errbuf);
+            // We need error object, not error value (JrS doesn't allow to
+            // pass the latter as a func argument).
+            jerry_value_clear_error_flag(&args[0]);
+        }
+
+        zjs_callback_id id = zjs_add_callback_once((jerry_value_t)user_data,
+                                      ZJS_UNDEFINED,
+                                      NULL,
+                                      NULL);
+        zjs_signal_callback(id, args, sizeof(args));
+        jerry_release_value(args[0]);
+    }
+}
+
+static jerry_value_t zjs_dgram_sock_send(const jerry_value_t function_obj,
+                                         const jerry_value_t this,
+                                         const jerry_value_t argv[],
+                                         const jerry_length_t argc)
+{
+    int ret;
+
+    if (argc < 5 || argc > 6)
```

Comment:
Again we normally only check for the right minimum, and the right types - pending discussion.

Probably should check for argv[0] being object, 1-2 being numbers, etc. I guess zjs_buffer_find on the wrong type would give you NULL like you're checking for, maybe that can be skipped.